content/blog/2022-05-19-A_Type_and_Effect_System_for_Deterministic_Parallel_Java/index.md

```diff
@@ -0,0 +1,123 @@
++++
+title = "A Type and Effect System for Deterministic Parallel Java"
+[extra]
+bio = """
+  [Nikita Lazarev](https://www.nikita.tech/) is an PhD student at Cornell's ECE department;
+  Susan Garry is an undergraduate student at Cornell's CS department.
+"""
+[[extra.authors]]
+name = "Nikita Lazarev"
+link = "https://www.nikita.tech/"  # Links are optional.
+[[extra.authors]]
+name = " Susan Garry"
++++
+
+### Paper review
+
+["A Type and Effect System for Deterministic Parallel Java", Robert L. Bocchino, Vikram S. Adve, Danny Dig, Sarita V. Adve, Stephen Heumann, Rakesh Komuravelli, Jeffrey Overbey, Patrick Simmons, Hyojin Sung, and Mohsen Vakilian. OOPSLA 2009](http://dpj.cs.illinois.edu/DPJ/Publications_files/DPJ-OOPSLA-2009.pdf)
+
+### Introduction
+
+Multithreading is notoriously difficult to reason about and error-prone. This happens due to potential race conditions that can be arise when two threads try to access the same region in memory simultaneously. Traditional approaches using locks (e.g. mutexes) can cause a deadlock and are also difficult to reason about and debug.
+
+In this paper, the authors propose making concurrency in Java safe and lock-free by using a type and effect system that involves partitioning and annotating Java programs. They introduce novel mechanisms that expand on the capabilities of previous type and effect systems, allowing for concurrent operations on recursive data structures via **region nesting**, concurrent array operations via **index-parameterized arrays** and **subarrays**, and concurrent writes to the same location in memory via **commutativity annotations**. They develop a new extension for Java called DPJ (Deterministic Parallel Java) which additionally comes with the formal verification functionality. The paper demonstrates how DPJ can be used to implement realistic algorithms and verify their thread-safeness.
```

Comment:
This paragraph starts by talking about safety in general, but that's not what makes this paper different: the focus is first and foremost on *determinism*. Everything else flows from that. To the extent that they avoid data races or other safety properties, it is *only* because it is necessary to provide determinism. So it is a somewhat bold statement saying that, if only parallel programs could be made deterministic, they would be easier to understand. You would still have other problems, but at least you would not have the problem of running the same program twice and getting different results. I think it's important to orient the technical discussion based on that fundamental goal.